odl/deform/linearized.py

```diff
@@ -0,0 +1,325 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Operators and functions for linearized deformation."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+from odl.operator.operator import Operator
+from odl import ProductSpace, DiscreteLp, Gradient, Divergence
+from odl import PointwiseInner, DiscreteLpVector
+import numpy as np
+
+
+__all__ = ('LinDeformFixedTempl', 'LinDeformFixedDisp')
+
+
+def _linear_deform(template, displacement, out=None):
+    """Linearized deformation of a template with a displacement field.
+
+    The function maps a gived template ``I`` and a given displacement
+    field ``v`` to the new function ``x --> I(x + v(x))``.
+
+    Parameters
+    ----------
+    template : `DiscreteLpVector`
+        Template to be deformed by a displacement field.
+    displacement : `ProductSpace` element
+        The vector field (displacement field) used in the linearized
+        deformation.
+
+    Returns
+    -------
+    deformed_template : `numpy.ndarray`
+
+    Examples
+    --------
+    >>> import odl
+    >>> space = odl.uniform_discr(0, 1, 5)
+    >>> disp_field_space = odl.ProductSpace(space, space.ndim)
+    >>> template = space.element([0, 0, 1, 0, 0])
+    >>> displacement_field = disp_field_space.element([[0, 0, 0, -0.2, 0]])
+    >>> _linear_deform(template, displacement_field)
+    array([ 0.,  0.,  1.,  1.,  0.])
+    """
+    image_pts = template.space.points()
+    for i, vi in enumerate(displacement):
+        image_pts[:, i] += vi.asarray().ravel()
+    return template.interpolation(image_pts.T, out=out, bounds_check=False)
+
+
+class LinDeformFixedTempl(Operator):
+
+    """Operator mapping a displacement field to corresponding deformed template.
+
+    The operator has a fixed template ``I`` and maps a displacement
+    field ``v`` to the new function ``x --> I(x + v(x))``.
+    """
+
+    def __init__(self, template, domain=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        template : `DiscreteLpVector` or element-like
+            Fixed template that is to be deformed.
```

Comment:
The should be adjusted so it fits inside the 80 chars, currently it is too narrow